source/library/Argo/Codec.hs

```diff
@@ -0,0 +1,284 @@
+arrayCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Array y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Array but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Array x
+        pure x
+    }
+
+objectCodec :: ValueCodec (Object.ObjectOf Value.Value)
+objectCodec = Codec
+    { decode = do
+        x <- Reader.ask
+        case x of
+            Value.Object y -> pure y
+            _ -> Trans.lift . Except.throwE $ "expected Object but got " <> show x
+    , encode = \ x -> do
+        State.put $ Value.Object x
+        pure x
+    }
+
+boolCodec :: ValueCodec Bool
+boolCodec = dimap Boolean.toBool Boolean.fromBool booleanCodec
+
+textCodec :: ValueCodec Text.Text
+textCodec = dimap String.toText String.fromText stringCodec
+
+eitherCodec :: ValueCodec a -> ValueCodec b -> ValueCodec (Either a b)
+eitherCodec c1 c2 = Codec
+    { decode = fmap Left (decode c1) <|> fmap Right (decode c2)
+    , encode = \ x -> case x of
+        Left y -> fmap Left $ encode c1 y
+        Right y -> fmap Right $ encode c2 y
+    }
+
+maybeCodec :: ValueCodec a -> ValueCodec (Maybe a)
+maybeCodec =
+    dimap
+        (either Just $ const Nothing)
+        (maybe (Right $ Null.fromUnit ()) Left)
+    . flip eitherCodec nullCodec
+
+data Permission
+    = Allow
+    | Forbid
+    deriving (Eq, Show)
+
+type ArrayCodec a = Codec
+    (State.StateT [Value.Value] (Except.Except String))
+    (Writer.Writer [Value.Value])
+    a
+
+fromArrayCodec :: Permission -> ArrayCodec a -> ValueCodec a
+fromArrayCodec p c = Codec
+    { decode = do
+        xs <- decode arrayCodec
+        case Except.runExcept . State.runStateT (decode c) $ Array.toList xs of
+            Left x -> Trans.lift $ Except.throwE x
+            Right (x, ys) -> do
+                case (p, ys) of
+                    (Forbid, _ : _) -> Trans.lift $ Except.throwE "leftover elements"
+                    _ -> pure ()
+                pure x
+    , encode = \ x -> do
+        Monad.void . encode arrayCodec . Array.fromList . Writer.execWriter $ encode c x
+        pure x
+    }
+
+element :: ValueCodec a -> ArrayCodec a
+element c = Codec
+    { decode = do
+        l <- State.get
+        case l of
+            [] -> Trans.lift $ Except.throwE "unexpected empty list"
+            h : t ->  case decodeWith c h of
+                Result.Failure y -> Trans.lift $ Except.throwE y
+                Result.Success y -> do
+                    State.put t
+                    pure y
+    , encode = \ x -> do
+        Writer.tell [encodeWith c x]
+        pure x
+    }
+
+tupleCodec :: ValueCodec a -> ValueCodec b -> ValueCodec (a, b)
+tupleCodec cx cy = fromArrayCodec Forbid $ (,)
+    <$> project fst (element cx)
+    <*> project snd (element cy)
+
+type ObjectCodec a = Codec
+    (State.StateT [Member.MemberOf Value.Value] (Except.Except String))
+    (Writer.Writer [Member.MemberOf Value.Value])
+    a
```

Comment:
There's a lot in common between this and the `ArrayCodec`. I wonder if both can lean on some more general "list codec"?